samples/matter/lock/src/software_images_swapper.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "software_images_swapper.h"
+
+#include <pm_config.h>
+
+#include <dfu/dfu_multi_image.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <zephyr/sys/reboot.h>
+
+#include <app/server/Server.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_MATTER_LOG_LEVEL);
+
+bool SoftwareImagesSwapper::Swap(ApplicationImageId applicationId)
+{
+	uint32_t app_address;
+	uint32_t app_size;
+	uint32_t net_address;
+	uint32_t net_size;
+
+	if (applicationId == ApplicationImageId::Image_1) {
+		app_address = PM_APP_1_CORE_APP_ADDRESS;
+		app_size = PM_APP_1_CORE_APP_SIZE;
+		net_address = PM_APP_1_CORE_NET_ADDRESS;
+		net_size = PM_APP_1_CORE_NET_SIZE;
+	} else if (applicationId == ApplicationImageId::Image_2) {
+		app_address = PM_APP_2_CORE_APP_ADDRESS;
+		app_size = PM_APP_2_CORE_APP_SIZE;
+		net_address = PM_APP_2_CORE_NET_ADDRESS;
+		net_size = PM_APP_2_CORE_NET_SIZE;
+	} else {
+		LOG_ERR("Requested to swap application image with invalid identifier: %d",
+			static_cast<uint8_t>(applicationId));
+		return false;
+	}
+
+	if (dfu_target_mcuboot_set_buf(mDFUImageBuffer, sizeof(mDFUImageBuffer)) != 0) {
+		return false;
+	}
+
+	if (!SwapImage(app_address, app_size, 0)) {
+		return false;
+	}
+
+	if (!SwapImage(net_address, net_size, 1)) {
+		return false;
+	}
+
+	if (dfu_target_schedule_update(-1) != 0) {
+		return false;
+	}
+
+	/* Wipe out whole settings as they will not apply to the new application image. */
+	chip::Server::GetInstance().ScheduleFactoryReset();
```

Comment:
I don't have a strong opinion on that, but maybe this call should be provided by the swapper API client as a callback. It would be more generic solution then, an we can avoid chip dependencies in the module itself.

Reply:
Yeah I thought about it, but in such case the application would need to care about doing factory reset after swap, what is a little bit counter intuitive. I was inspired by the OTA DFU module that triggers reset without involving an app. I don't have strong opinion either, so we may wait for others opinion on that.

Reply:
I added callback, as suggested.